### PR TITLE
feat: initial message signing implementation

### PR DESCRIPTION
**Pull request by obmarg**
This PR implements the majority of message signing.  We now take the config, and apply a signature to any headers that the user has requested we sign.  We do not yet support any derived components or signature parameters or remove the hop-by-hop headers - those are still to be done.

Fixes GB-7685
Fixes GB-7787
Fixes GB-7864

### CODE REVIEW

File: engine/crates/integration-tests/src/federation/builder.rs [hunk 1]
```diff
@@ -27,7 +28,11 @@ pub struct TestGatewayBuilder {
     mock_subgraphs: Vec<(TypeId, String, BoxFuture<'static, MockGraphQlServer>)>,
     docker_subgraphs: HashSet<DockerSubgraph>,
     config_source: Option<ConfigSource>,
-    runtime: TestRuntime,
+
+    trusted_documents: Option<trusted_documents_client::Client>,
+    hooks: Option<DynamicHooks>,
+    fetcher: Option<DynamicFetcher>,
+    // runtime: TestRuntime,
```
Comment (obmarg) [inline engine/crates/integration-tests/src/federation/builder.rs:35]:
```suggestion
    fetcher: Option<DynamicFetcher>,
```

File: engine/crates/runtime-local/src/fetch/signing.rs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            return Ok(None);
+        }
+
+        let Some(key) = key else { todo!("error handling") };
+        let key_id = key.id().map(str::to_string);
+        let key = Key::load_from(&key, algorithm)?;
+
+        // TODO: Add the hop-by-hop headers into exclude headers by default...
+        // TODO: Also add the default derived_components and/or signature_parameters
+
```
Comment (obmarg) [inline engine/crates/runtime-local/src/fetch/signing.rs:99]: Doing these in my next PR

File: engine/crates/runtime-local/src/fetch/signing.rs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    fn httpsig_params(&self, headers: &http::HeaderMap) -> anyhow::Result<HttpSignatureParams> {
+        let covered_components = headers
+            .iter()
+            .filter(|(name, _)| self.should_include_header(name.as_str()))
+            .map(|(name, _)| HttpMessageComponentId::try_from(name.as_str()))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // TODO: Handle derived_components & signature parameters
```
Comment (obmarg) [inline engine/crates/runtime-local/src/fetch/signing.rs:118]: Also going to do this in next PR

File: engine/crates/runtime-local/src/fetch/signing.rs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+        let mut params = HttpSignatureParams::try_new(&covered_components)?;
+        params.set_key_info(&self.key);
+        if let Some(id) = &self.key_id {
+            params.set_keyid(id);
+        }
+
+        if let Some(expiry) = self.expiry {
+            // TODO: Ok, so this sets it from now, need to do something else...
+            params.set_expires_with_duration(Some(expiry.as_secs()));
```
Comment (obmarg) [inline engine/crates/runtime-local/src/fetch/signing.rs:128]:
```suggestion
            params.set_expires_with_duration(Some(expiry.as_secs()));
```

**Review by pimeys — APPROVED** [review]

**Pull request merged**